opc/data_source_machine_image.go

```diff
@@ -0,0 +1,134 @@
+package opc
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/go-oracle-terraform/compute"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/structure"
+)
+
+func dataSourceMachineImage() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceMachineImageRead,
+
+		Schema: map[string]*schema.Schema{
+			"account": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"attributes": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"description": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"error_reason": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"hypervisor": {
+				Type:     schema.TypeMap,
+				Computed: true,
+			},
+
+			"image_format": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"file": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"no_upload": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+
+			"platform": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"sizes": {
+				Type:     schema.TypeMap,
+				Computed: true,
+			},
+
+			"state": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"uri": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceMachineImageRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*OPCClient).computeClient.MachineImages()
+
+	// Get required attributes
+	account := d.Get("account").(string)
+	name := d.Get("name").(string)
+
+	input := compute.GetMachineImageInput{
+		Account: account,
+		Name:    name,
+	}
+
+	result, err := client.GetMachineImage(&input)
+	if err != nil {
+		return err
+	}
+
+	// Not found, don't error
+	if result == nil {
+		d.SetId("")
+		return nil
+	}
+
+	// Flatten JSON attributes
+	attributes, err := structure.FlattenJsonToString(result.Attributes)
+	if err != nil {
+		return err
+	}
+
+	// Populate schema attributes
+	d.SetId(fmt.Sprintf("%s", result.Name)) // TODO
```

Comment:
Is there a TODO here or is this a leftover comment?

Reply:
Ah yes, I wasn't entirely sure what should be set for `SetId()` - for the other data source I looked at for examples they are concatenating multiple attributes into the SetId() - for the machine image name should be enough the be unique - but I did wonder if the Id should be set to the fully qualified name e.g. `/Compute-mydomain/user@example.com/my-machine-image`

Reply:
The name should be fine as an ID here. We had talked about doing fqdn for resources and I think this would be a good starting resource if you wanted to add that here.

Reply:
I've left it as just name, we should think though how we consistently handle the fully qualified resource names as apply across all resources in a single update, for another time...